### PR TITLE
Show Ghost Upgrade Tiers

### PR DESCRIPTION
**Pull request by trevor-sutherland**
[…estinyItemManager/DIM/issues/8664](https://github.com/DestinyItemManager/DIM/issues/8664)

### CODE REVIEW

File: src/app/inventory/store/energy.ts [hunk 1]
```diff
@@ -26,9 +26,16 @@ export function energyUpgrade(
...
   newEnergyCapacity: number
 ) {
-  const tierSocket =
+  let tierSocket;
+  const armorTierSocket =
     item.sockets && getFirstSocketByCategoryHash(item.sockets, SocketCategoryHashes.ArmorTier);
-
+  const ghostTierSocket =
+    item.sockets && getFirstSocketByCategoryHash(item.sockets, SocketCategoryHashes.GhostTier);
+  tierSocket = armorTierSocket;
```
Comment (delphiactual) [inline src/app/inventory/store/energy.ts:34]:
```suggestion
  const tierSocket = armorTierSocket?.plugSet ? armorTierSocket : ghostTierSocket;
```

File: src/app/inventory/store/energy.ts [hunk 2]
```diff
@@ -26,9 +26,16 @@ export function energyUpgrade(
   newEnergyType: DestinyEnergyType,
   newEnergyCapacity: number
 ) {
-  const tierSocket =
+  let tierSocket;
```
Comment (delphiactual) [inline src/app/inventory/store/energy.ts:29]: remove

File: src/app/inventory/store/energy.ts [hunk 3]
```diff
@@ -26,9 +26,16 @@ export function energyUpgrade(
...
+  const armorTierSocket =
     item.sockets && getFirstSocketByCategoryHash(item.sockets, SocketCategoryHashes.ArmorTier);
-
+  const ghostTierSocket =
+    item.sockets && getFirstSocketByCategoryHash(item.sockets, SocketCategoryHashes.GhostTier);
+  const tierSocket = armorTierSocket?.plugSet ? armorTierSocket : ghostTierSocket;
+  // We want to display energy info for ghosts
+  if (!armorTierSocket?.plugSet) {
+    tierSocket = ghostTierSocket;
+  }
```
Comment (delphiactual) [inline src/app/inventory/store/energy.ts:38]: remove

**Review by delphiactual — APPROVED** [review]

**Comment by delphiactual** [comment]
```
  const tierSocket =
    item.sockets &&
    (getFirstSocketByCategoryHash(item.sockets, SocketCategoryHashes.ArmorTier) ||
      getFirstSocketByCategoryHash(item.sockets, SocketCategoryHashes.GhostTier));
```

File: src/app/inventory/store/energy.ts (hunk 2 above)
Comment (delphiactual) [inline src/app/inventory/store/energy.ts:29]:
```suggestion
  const tierSocket =
    item.sockets &&
    (getFirstSocketByCategoryHash(item.sockets, SocketCategoryHashes.ArmorTier) ||
      getFirstSocketByCategoryHash(item.sockets, SocketCategoryHashes.GhostTier));
```

File: src/app/inventory/store/energy.ts [hunk 4]
```diff
@@ -26,9 +26,16 @@ export function energyUpgrade(
   newEnergyType: DestinyEnergyType,
   newEnergyCapacity: number
 ) {
-  const tierSocket =
+  let tierSocket;
+  const armorTierSocket =
```
Comment (delphiactual) [inline src/app/inventory/store/energy.ts:30]: remove

File: src/app/inventory/store/energy.ts (hunk 3 above)
Comment (delphiactual) [inline src/app/inventory/store/energy.ts:38]: remove

**Review by delphiactual — CHANGES_REQUESTED** [review]

File: src/app/item-popup/EnergyMeter.tsx [hunk 5]
```diff
@@ -202,7 +204,12 @@ function EnergyUpgradePreview({
   previewEnergyType: DestinyEnergyType;
 }) {
   const defs = useD2Definitions()!;
-  if (!item.energy || !swappableEnergyTypes.includes(item.energy.energyType)) {
+  // return null if not swappable and not a ghost
+  if (
+    !item.energy ||
+    (!swappableEnergyTypes.includes(item.energy.energyType) &&
+      !permanentEnergyTypes.includes(item.energy.energyType))
+  ) {
```
Comment (delphiactual) [inline src/app/item-popup/EnergyMeter.tsx:212]:
```suggestion
  // return null if not swappable and not a ghost
  if (
    !item.energy ||
    ![...swappableEnergyTypes, DestinyEnergyType.Ghost].includes(item.energy.energyType) {
```

File: src/app/item-popup/EnergyMeter.tsx [hunk 6]
```diff
@@ -27,6 +27,8 @@ const swappableEnergyTypes = [
   DestinyEnergyType.Stasis,
 ];
 
+const permanentEnergyTypes = [DestinyEnergyType.Ghost];
+
```
Comment (delphiactual) [inline src/app/item-popup/EnergyMeter.tsx:31]: remove

**Review by delphiactual — CHANGES_REQUESTED** [review]

**Review by delphiactual — APPROVED** [review]

File: src/app/dim-ui/ConfirmButton.m.scss [hunk 7]
```diff
@@ -36,7 +36,7 @@
 
     &[class~='danger'] {
       &:hover {
-        background-color: #aa2222 !important;
+        background-color: #a22 !important;
```
[the remark this answers is not in the document]
Comment (trevor-sutherland) [reply src/app/dim-ui/ConfirmButton.m.scss:39]: My bad, I didn't know how/why I committed that. "yarn fix" addition maybe?

**Pull request merged**